### PR TITLE
Add cmake files for cpp bindings

### PR DESCRIPTION
**Pull request by Febbe**

### CODE REVIEW

**Comment by Febbe** [comment]
This cmake extension for the cpp bindings is designed to stand alone. I could integrate it as Component for the main cmake-package, but as you want to split those it would be a waste of effort.

**Comment by carlo-bramini** [comment]
Hello, I would like to suggest to add something like this:

```
option( enable-cxx "Enable C++ bindings" off )
```

and then replace the fix into the main `CMakeLists.txt` with something like this:

```
if (enable-cxx)
  add_subdirectory(bindings/cpp)
endif(enable-cxx)
```

So, if one day the C++ bindings will be moved outside this project into a separate repository, it will be still possible to configure Portaudio. Instead, if they are included because you cloned its submodule, then you can also build it by adding a simple `-Denable-cxx=ON` when calling CMake.
What do you think?

**Comment by Febbe** [comment]
> Hello, I would like to suggest to add something like this:
> 
> ```
> option( enable-cxx "Enable C++ bindings" off )
> ```
> 
> and then replace the fix into the main `CMakeLists.txt` with something like this:
> 
> ```
> if (enable-cxx)
>   add_subdirectory(bindings/cpp)
> endif(enable-cxx)
> ```
> 
> So, if one day the C++ bindings will be moved outside this project into a separate repository, it will be still possible to configure Portaudio. Instead, if they are included because you cloned its submodule, then you can also build it by adding a simple `-Denable-cxx=ON` when calling CMake. What do you think?

Using something like EXCLUDE_FROM_ALL might be better.

**Comment by RossBencina** [comment]
Given that we're definitely going to move the C++ binding out of the main repo I don't think it makes sense to add it to the main `CMakeListstxt` in the interim.

**Comment by RossBencina** [comment]
Please remove the change to the main `CMakeLists.txt` and we can merge this.

**Comment by RossBencina** [comment]
Looks good. I've assigned to @philburk for review

**Comment by RossBencina** [comment]
@Be-ing is this PR OK by you?

**Review by RossBencina — APPROVED** [review]

File: bindings/cpp/CMakeLists.txt
Comment (Be-ing) [inline bindings/cpp/CMakeLists.txt:47]: https://cmake.org/cmake/help/latest/command/file.html#glob-recurse
> We do not recommend using GLOB to collect a list of source files from your source tree. If no CMakeLists.txt file changes when a source is added or removed then the generated build system cannot know when to ask CMake to regenerate. The CONFIGURE_DEPENDS flag may not work reliably on all generators, or if a new generator is added in the future that cannot support it, projects using it will be stuck. Even if CONFIGURE_DEPENDS works reliably, there is still a cost to perform the check on every rebuild.

File: bindings/cpp/CMakeLists.txt
Comment (Be-ing) [inline bindings/cpp/CMakeLists.txt:5]: These C++ files haven't even been substantially modified since the first Git commit in 2008.

File: bindings/cpp/CMakeLists.txt
Comment (Be-ing) [inline bindings/cpp/CMakeLists.txt:11]: IMO this indirection makes this harder to read for no clear benefit. I would prefer the list of source files to be in this file, not delegated to a subdirectory. You can simply list them in the `add_library` call below.

File: bindings/cpp/source/portaudiocpp/CMakeLists.txt
Comment (Be-ing) [inline bindings/cpp/source/portaudiocpp/CMakeLists.txt:9]:
```suggestion
    BlockingStream.cxx
```
unneeded when listed in parent directory's CMakeLists.txt

File: bindings/cpp/CMakeLists.txt
Comment (Be-ing) [inline bindings/cpp/CMakeLists.txt:32]: Remove this line. It reduces compatibility for no reason.

File: bindings/cpp/CMakeLists.txt
Comment (Be-ing) [inline bindings/cpp/CMakeLists.txt:29]:
```suggestion
    $<INSTALL_INTERFACE:${CMAKE_INSTALL_INCLUDEDIR}>
```
I think GNUInstallDirs needs to be included before this.

File: bindings/cpp/CMakeLists.txt
Comment (Be-ing) [inline bindings/cpp/CMakeLists.txt:16]: This comment is confusing. This is portaudio-cpp's CMakeLists.txt, so why is this still here? Do you mean remove this when the C++ bindings are moved to another repository?

**Review by Be-ing — CHANGES_REQUESTED** [review]
SOVERSION is missing, see https://github.com/PortAudio/portaudio/blob/06bc8763e72b6aa5fa1f380946e09b676e09806f/CMakeLists.txt#L385

**Comment by RossBencina** [comment]
@Febbe please respond to @Be-ing's review.

@Be-ing thanks.

File: bindings/cpp/CMakeLists.txt
Comment (Febbe) [reply bindings/cpp/CMakeLists.txt:5]: I've added this with the possibility in mind to modernize the code. 
The code should be at least modernized to c++11 and since there are no relevant breaking changes between 11 and 14 but enourmous improvements, it is better to skip c++11.

Another sidefact, llvm is based on cpp14 and msvc does not even support c++11 anymore. The minimal supported version is c++14.

File: bindings/cpp/CMakeLists.txt
Comment (Febbe) [reply bindings/cpp/CMakeLists.txt:29]: No, GNUInstallDirs is only required for the GNUInstallDir's variables.

`$<INSTALL_INTERFACE:include>` means, that the includes can be found in an subdirectory of the folder `include` (which is an relative path). Therefore it will be `$install_dir/include`.

File: bindings/cpp/CMakeLists.txt
Comment (Febbe) [reply bindings/cpp/CMakeLists.txt:47]: This is only relevant, when a file in the GLOB is removed, added, or moved. This is often not relevant for developers, because they often call cmake __before every__ build.

In this case it is irrelevant, because those files are collected only for installation.

File: bindings/cpp/CMakeLists.txt
Comment (Febbe) [reply bindings/cpp/CMakeLists.txt:16]: Yep, forgot that line.

File: bindings/cpp/CMakeLists.txt
Comment (Febbe) [reply bindings/cpp/CMakeLists.txt:11]: Mhh, I've done this, to reduce the bloat in the main cmake file. For the understanding what this code does, it is not relevant, which source file is added, but it is required to add them manually. (To prevent [this](https://cmake.org/cmake/help/latest/command/file.html#glob-recurse)) 

it might be more readable to move the contents in a add_sources.cmake file, instead an subdirectory. Is this better for you, or should I still remove it?

**Comment by carlo-bramini** [comment]
I tried to test this patch but I was unable to make it working.
Actually, CMake emits these errors:

```
  Could not find a package configuration file provided by "ASIO" with any of
  the following names:

    ASIOConfig.cmake
    asio-config.cmake
```

I have been able to solve the trouble by doing these actions:
1) I created a "modules" directory into "cmake" and I copied FindASIO.cmake from the main sources of Portaudio.
2) Next, I added this line:

```
set(CMAKE_MODULE_PATH "${CMAKE_MODULE_PATH}" "${CMAKE_CURRENT_SOURCE_DIR}/cmake/modules")
```

to CMakeLists.txt.
Next, I got this error:

```
CMake Error at cmake/modules/FindASIO.cmake:20 (message):
  ASIO is only supported on Windows.
Call Stack (most recent call first):
  source/portaudiocpp/CMakeLists.txt:1 (find_package)
```

I solved the problem by doing an if `(WIN32) ... endif()` inside `bindings\cpp\source\portaudiocpp\CMakeLists.txt`.
Then, I got again another error:

```
CMake Error at CMakeLists.txt:21 (add_library):
  add_library cannot create ALIAS target "PortAudio::portaudio" because
  target "portaudio" does not already exist.
```

and here I stopped because I did not know exactly what to do.
I hope that this report could be useful.

**Comment by Be-ing** [comment]
@carlo-bramini Did you try to run CMake from the bindings/cpp subdirectory? This CMakeLists.txt is not meant for that. Run CMake on the root of the portaudio repository.

File: bindings/cpp/CMakeLists.txt
Comment (Be-ing) [reply bindings/cpp/CMakeLists.txt]: Modernizing the code sounds like a good idea, but until that's done, there's no reason to reduce compatibility.

File: bindings/cpp/CMakeLists.txt
Comment (Be-ing) [reply bindings/cpp/CMakeLists.txt:44]: > This is often not relevant for developers

It's not relevant until it is relevant and someone gets confused. Please list the files individually; it is more robust.

File: bindings/cpp/CMakeLists.txt
Comment (Be-ing) [reply bindings/cpp/CMakeLists.txt:11]: Moving this to a separate module wouldn't help anything.

**Comment by carlo-bramini** [comment]
> @carlo-bramini Did you try to run CMake from the bindings/cpp subdirectory? This CMakeLists.txt is not meant for that. Run CMake on the root of the portaudio repository.

Yes, I pointed CMake to bindings/cpp directory.
Since `add_subdirectory(bindings/cpp)` has been removed from the parent CMakeLists.txt at least from my last checkout, it will never work in other ways.
Running CMake from the root of the portaudio repository just builds libportaudio and nothing else on my PC.

**Comment by Febbe** [comment]
> Actually, CMake emits these errors:

The cmake errors regarding Asio are no errors. Actually, they are warnings.
To configure this project, asio is not required. But on linux, where asio should be included already, this should not be required at all. I'll take a look into it.

> ``` 
  CMake Error at CMakeLists.txt:21 (add_library):
  add_library cannot create ALIAS target "PortAudio::portaudio" because
  target "portaudio" does not already exist.
```
and here I stopped because I did not know exactly what to do.
I hope that this report could be useful.

Do you know, which targets are found by find_package(Portaudio)?
Since we don't call the CMakeLists.txt of portaudiocpp inside the CMakeLists.txt of portaudio, it is mandatory, that portaudio is either installed or configured before portaudiocpp is configured.

File: bindings/cpp/CMakeLists.txt
Comment (Febbe) [reply bindings/cpp/CMakeLists.txt]: Done, but I disagree in this case, developers usually don't control, if a header is in the install set. Therefore, it would be more robust the other way round.

**Comment by Febbe** [comment]
@carlo-bramini your issues should be fixed now. 
@Be-ing your change requests are implemented, but please don't merge this yet. I am on vacation, and therefore I don't have an windows machine to test this.

**Comment by carlo-bramini** [comment]
> @carlo-bramini your issues should be fixed now.

I tested it with CYGWIN and MinGW-w64 and it worked fine.
Running CMake-GUI and selecting MSVC compiler, it prints these error messages:

```
The CXX compiler identification is MSVC 19.29.30146.0
Detecting CXX compiler ABI info
Detecting CXX compiler ABI info - done
Check for working CXX compiler: C:/Program Files (x86)/Microsoft Visual Studio/2019/Community/VC/Tools/MSVC/14.29.30133/bin/Hostx64/x64/cl.exe - skipped
Detecting CXX compile features
Detecting CXX compile features - done
ASIO SDK NOT found
Could NOT find PkgConfig (missing: PKG_CONFIG_EXECUTABLE) 
CMake Error at C:/Program Files/CMake/share/cmake-3.22/Modules/FindPkgConfig.cmake:659 (message):
  pkg-config tool not found
Call Stack (most recent call first):
  C:/Program Files/CMake/share/cmake-3.22/Modules/FindPkgConfig.cmake:825 (_pkg_check_modules_internal)
  cmake/modules/FindPortAudio.cmake:27 (pkg_check_modules)
  CMakeLists.txt:73 (find_package)

CMake Error at cmake/modules/FindPortAudio.cmake:28 (add_library):
  add_library cannot create ALIAS target "PortAudio::portaudio" because
  target "PkgConfig::portaudio" does not already exist.
Call Stack (most recent call first):
  CMakeLists.txt:73 (find_package)

Configuring incomplete, errors occurred!
```

**Comment by Febbe** [comment]
Ah ok, I can make pkg-config optional, but then a port audio installation, which is available via pkg-config cant be found. This means, that portaudio must be installed via cmake, and the path to it must be known.

**Comment by Febbe** [comment]
@carlo-bramini you have to compile portaudio first and install it. Then you have to set the PortAudio_DIR env variable.

I would add a fetch_content to pull portaudio, but this should be done, after those 2 projects are split. It is also pulled already.

@Be-ing why you don't create subprojects of your PortAudio group like 
`PortAudio/portaudiojava` and
`PortAudio/portaudiocpp`

**Comment by RossBencina** [comment]
Regarding ASIO detection, this might be relevant: https://github.com/PortAudio/portaudio/issues/518

**Comment by philburk** [comment]
@Febbe wrote:
> please don't merge this yet. I am on vacation,

Please let us know when you are back from vacation and have had a chance to test the recent changes.
Any work remaining?

**Comment by Febbe** [comment]
> @Febbe wrote:
> 
> > please don't merge this yet. I am on vacation,
> 
> Please let us know when you are back from vacation and have had a chance to test the recent changes. Any work remaining?

Thanks for the reminder, I've tested it, and it compiles on both ubuntu and windows 10

**Comment by Febbe** [comment]
@philburk I've squashed and rebased everything into one commit on top current master. Should be mergeable now.

**Comment by RossBencina** [comment]
I notice that your `FindASIO.cmake` is not the same as the one in `/portaudio/cmake/modules`. I would like them to match. For now, please either change the master one, or revert your changes, or use the file in `/portaudio/cmake/modules` by adding a relative path to `/portaudio/cmake/modules` to `CMakeLists.txt`. 

When the cpp bindings are split to their own repository then a decision whether to use the portaudio copy via a submodule or maintain a divergent copy.

Note that there are pending issues with `FindASIO.cmake` so it is likely to change.

**Comment by philburk** [comment]
@Be-ing - Do these final changes look good to you? Were your comments addressed?
We would like to get your approval before merging.

File: bindings/cpp/CMakeLists.txt
Comment (daschuer) [inline bindings/cpp/CMakeLists.txt:6]: This is not required. I can successfully build without. As a general purpose library the requirements should be as low as possible.

**Comment by RossBencina** [comment]
My comment above regarding `FindASIO.cmake` has not been addressed.

https://github.com/PortAudio/portaudio/pull/721#issuecomment-1306253950

**Comment by RossBencina** [comment]
`FindASIO.cmake` still do not match entirely, but I'm not going to quibble about this minor change.

```
if(NOT WIN32)
  message(FATAL_ERROR "ASIO is only supported on Windows.")
endif()
```
vs
```
if(NOT WIN32)
  message(WARNING "ASIO is only supported on Windows.")
  set(ASIO_FOUND OFF)
  return()
endif()
```

File: cmake/modules/FindASIO.cmake
Comment (RossBencina) [inline cmake/modules/FindASIO.cmake:22]: Contrary to my previous (now retracted) message. This change to the PortAudio library CMake behavior does not really belong in a PR that is supposed to be specific to the C++ library. If this change is generally considered to be desirable it should be in a separate PR. Suggest rolling it back:

```suggestion
  message(FATAL_ERROR "ASIO is only supported on Windows.")
```

**Comment by RossBencina** [comment]
@Be-ing @daschuer @dechamps and others, do you have any final comments before we merge this?

**Comment by daschuer** [comment]
I have no additional comments.

File: bindings/cpp/portaudiocpp.pc.in
Comment (Febbe) [reply bindings/cpp/portaudiocpp.pc.in:4]: Oh, you may be right. It's better, to copy the file into the cmake folder then.

Comment (Febbe) [reply bindings/cpp/portaudiocpp.pc.in]: I copied the file to `cmake/portaudiocpp.pc.in` and reverted the changes to the original config.

**Review by philburk — APPROVED** [review]

**Comment by philburk** [comment]
Some time in the future, it would be nice to have a CI target that tests the C++ build.

**Comment by RossBencina** [comment]
Thankyou!

**Pull request merged**
